src/include/storage/index/bplustree.h

```diff
@@ -747,6 +749,501 @@ class BPlusTree : public BPlusTreeBase {
+
+        // Get shared latch on current node, and release the prev_node.
+        if (!(curr_node_->TrySharedLock())) {
+          prev_node->ReleaseNodeSharedLatch();
+          SetRetryIterator();
+          return;
+        }
+        prev_node->ReleaseNodeSharedLatch();
+
+        curr_key_ = curr_node_->Begin();
+        curr_val_ = curr_key_->second->begin();
+      }
+
+      first_ = curr_key_->first;
+      second_ = *curr_val_;
+    }
+
+    /**
+     * Prefix Decrement operator to perform backward iteration
+     */
+    void operator--() {
+      NOISEPAGE_ASSERT(state_ == VALID, "Iterator in Invalid State.");
+
+      // Move by one value
+      curr_val_++;
+      if (curr_val_ == curr_key_->second->end()) {
+        // If the current list has been exhausted, move to the next key
+        curr_key_--;
+        if (curr_key_ != curr_node_->REnd()) {
+          curr_val_ = curr_key_->second->begin();
+        }
+      }
+
+      if (curr_key_ == curr_node_->REnd()) {
+        // If the current node has been exhausted, move to the next node
+        if (curr_node_->GetLowKeyPair().second == nullptr) {
+          curr_node_->ReleaseNodeSharedLatch();
+          SetREndIterator();
+          return;
+        }
+        auto prev_node = curr_node_;
+        curr_node_ = reinterpret_cast<ElasticNode<KeyValuePair> *>(curr_node_->GetLowKeyPair().second);
+
+        // Get shared latch on current node, and release the prev_node.
+        if (!(curr_node_->TrySharedLock())) {
+          prev_node->ReleaseNodeSharedLatch();
+          SetRetryIterator();
+          return;
+        }
+        prev_node->ReleaseNodeSharedLatch();
+
+        curr_key_ = curr_node_->RBegin();
+        curr_val_ = curr_key_->second->begin();
+      }
+
+      first_ = curr_key_->first;
+      second_ = *curr_val_;
+    }
+
+    /**
+     * Equality operator to check if two iterators are equal
+     * @param itr Iterator to be compared with this iterator
+     * @return true if iterators are equal, false otherwise
+     */
+    bool operator==(const BPlusTreeIterator &itr) {
+      bool result = (curr_node_ == itr.curr_node_ && curr_key_ == itr.curr_key_ && state_ == itr.state_);
+      if (state_ == VALID) {
+        result = result && (curr_val_ == itr.curr_val_);
+      }
+      return result;
+    }
+
+    /**
+     * Inequality operator to check if two iterators are inequal
+     * @param itr Iterator to be compared with this iterator
+     * @return true if iterators are inequal, false otherwise
+     */
+    bool operator!=(const BPlusTreeIterator &itr) {
+      bool result = (curr_node_ == itr.curr_node_ && curr_key_ == itr.curr_key_ && state_ == itr.state_);
+      if (state_ == VALID) {
+        result = result && (curr_val_ == itr.curr_val_);
+      }
+      return !result;
+    }
+
+    /**
+     * Function to notify that iteration is complete. This is used to release any pending
+     * locks that is being held
+     */
+    void Done() {
```

Comment:
I'm still not sold on it being in the destructor. We should chat in the standup.